new-frontiers-14/frontier-station-14
# Revolvers can reload from boxes and magazines.

## About the PR



With this PR, revolvers can take cartridges directly from boxes, magazines, or other revolvers.  You can also put bullets from revolvers directly into boxes, magazines, or speedloaders.  A fix was also made to speedloader reloads - partial reloads could previously result in cartridges disappearing.

## Why / Balance


Having to reload revolvers one cartridge at a time was fiddly.  Considering that shotguns and pistol magazines can be reloaded from boxes without issue, it seems reasonable that revolvers can also reload from boxes.

This should make the use of revolvers more common, which I would argue is fitting to a spaghetti western/frontier sort of game.  The reload rate (currently ~4.2s for a Python) and the ability for revolvers to reload from other revolvers can be tweaked, just as with BallisticAmmoProvider components.

## How to test


### Expected behaviour
* Revolvers can reload from ammo boxes, magazines, speedloaders, or other revolvers of the same caliber.
* Revolvers cannot reload from ammo boxes, magazines, speedloaders or revolvers of different caliber.
* Revolvers should still be able to load from individual cartridges of the same caliber.
* Revolvers should not be cycled when reloading something else or being reloaded.
* Spent rounds in a revolver will remain when reloading something else from the revolver.
* Previous behaviour of ammo boxes and magazines (reloading from other ammo boxes and magazines, individual cartridges) should not change.
* All rounds removed from a speedloader during a reload should be placed in the revolver receiving them.

Known bug: when reloading something with a revolver of an incorrect cartridge type, two messages show up - an invalid cartridge type message and _Inserted_, and there will be a bullet missing from the revolver's HUD.

### Example Test Procedure

Build and run local server.
1. Spawn two Pythons, an Argenti, a box of magnum ammunition, a box of .20 rifle ammunition, and any full magnum speedloader.
2. Empty the first Python.  Reload the second into the first.  First should have 6 cartridges when finished, second should have 0.
3. Fire a round in the first Python.  Reload the second from the first.  The first should have 1 spent cartridge, second should have 5 unspent cartridges.
4. Cycle the first Python twice and reload it from the magnum box.  The first Python should have 1 spent cartridge and 5 unspent cartridges, and the box should have 7 rounds in it.
5. Reload the box from the first Python.  The first Python should have 1 spent cartridge and the box should have 12 rounds in it.
6. Insert three of the .45 cartridges on the ground into the second Python.  Shoot one.  Reload the second Python from the speedloader.  The second Python should have 5 unspent cartridges and 1 spent cartridge, the speedloader should have three unspent rounds.
7. Reload the speedloader from the second Python.  The second Python should have 2 unspent cartridges and 1 spent cartridge, and the speedloader should have six unspent rounds.
8. Try to reload the second Python from the Argenti.  It should not work, the Argenti should still be full.
9. Try to reload the second Python from the box of .20 rifle ammunition.  It should not work, the box should still be full.

## Media




https://github.com/new-frontiers-14/frontier-station-14/assets/166147148/a2bdcbe7-6fcc-43ea-bd57-c64e530a1653

## Breaking changes


RevolverAmmoComponent now has two extra public fields: MayTransfer and FillDelay.
TakeAmmoEvent has a new public member, WillBeFired, and its constructor receives this as a parameter that defaults to false.

### Rationale

RevolverAmmoComponent now more closely mimics BallisticAmmoComponent, and these two fields were brought over to configure the new, similar behaviour.
The TakeAmmoEvent handler in the revolver previously was used only for firing rounds - any rounds removed would be fired and spent.  The difference of intent between removing a round to be fired and removing a round to be placed elsewhere could not be 

**Changelog**



:cl:
- add: Revolvers can now reload from ammunition boxes and magazines, and can provide ammunition for speedloaders.
- fix: Speedloaders do not waste ammunition on partial reloads.
